sleap_roots/graphpipeline.py

```diff
@@ -169,9 +188,287 @@ def get_traits_value_frame(
+            pts_all_array = lateral_pts
+            pts_all_list = list(lateral_pts)
+        # if no lateral root, just give two nan points
+        if get_root_lengths(lateral_pts).shape[0] == 0:
+            lateral_pts = np.array([[(np.nan, np.nan), (np.nan, np.nan)]])
+
+        data = get_traits_value_frame(
+            primary_pts, lateral_pts, pts_all_array, pts_all_list
+        )
+        data["plant_name"] = plant_name
+        data["frame_idx"] = frame
+        data_plant = pd.concat([data_plant, data], ignore_index=True)
+
+    # reorganize the column position
+    column_names = data_plant.columns.tolist()
+    column_names = [column_names[-2]] + [column_names[-1]] + column_names[:-2]
+    data_plant = data_plant[column_names]
+
+    if write_csv:
+        data_plant.to_csv("plant_original_traits.csv", index=False)
+    return data_plant
+
+
+def get_traits_value_plant_summary(
+    h5,
+    lateral_only: bool = False,
+    write_csv: bool = False,
+) -> pd.DataFrame:
+    """Get summarized SLEAP traits per plant based on graph.
+
+    Args:
+        h5: h5 file, plant image series.
+        lateral_only: Boolean value, where false is rice (default), true is dicot.
+        write_csv: Boolean value, where true is write csv file.
+
+    Return:
+        A DataFrame with all summarized traits per plant.
+    """
+    # lateral_only = False
+    # write_csv = False
+    data_plant = get_traits_value_plant(h5, lateral_only, write_csv)
+
+    scalar_traits = [
+        "primary_angle_proximal",
+        "primary_angle_distal",
+        "primary_length",
+        "primary_base_tip_dist",
+        "primary_depth",
+        "lateral_count",
+        "grav_index",
+        "base_length",
+        "base_length_ratio",
+        "primary_tip_pt_y",
+        "base_median_ratio",
+        "base_ct_density",
+        "chull_perimeter",
+        "chull_area",
+        "chull_max_width",
+        "chull_max_height",
+        "ellipse_a",
+        "ellipse_b",
+        "ellipse_ratio",
+        "network_width_depth_ratio",
+        "network_solidity",
+        "network_length_lower",
+        "network_distribution_ratio",
+        "scanline_first_ind",
+        "scanline_last_ind",
+    ]
+
+    non_scalar_traits = [
+        "lateral_angles_proximal",
+        "lateral_angles_distal",
+        "lateral_lengths",
+        "stem_widths",
+        "lateral_base_xs",
+        "lateral_base_ys",
+        "lateral_tip_xs",
+        "lateral_tip_ys",
+        "chull_line_lengths",
+        "scanline_intersection_counts",
+    ]
+
+    # get summarized non-scalar traits per frame
+    data_plant_frame_summary = data_plant[["plant_name", "frame_idx"]]
+
+    for i in range(len(non_scalar_traits)):
+        print(non_scalar_traits[i])
+        trait = data_plant[non_scalar_traits[i]]
+        for j in range(len(trait)):
+            if type(trait[j]) == np.ndarray and trait[j].shape == (trait[j].size,):
+                trait[j] = trait[j].reshape((1, -1))
```

Comment:
When does this happen?

Reply:
Those are reshape array of (n,) to (n,1). The purpose of doing this is to filter out [] non-scalar trait, since there is no statistical summary of [] arrays.